engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java

```diff
@@ -0,0 +1,108 @@
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.builtin.Builtins;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+import org.enso.interpreter.runtime.number.EnsoBigInteger;
+import org.enso.interpreter.runtime.type.TypesGen;
+
+@BuiltinMethod(
+    type = "Meta",
+    name = "type_of_builtin",
+    description = "Returns the type of a value.")
+public abstract class TypeOfNode extends Node {
+
+  abstract Object execute(@AcceptsError Object value);
+
+  static TypeOfNode build() {
+    return TypeOfNodeGen.create();
+  }
+
+  @Specialization
+  Object doDouble(double value) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().number().getDecimal();
+  }
+
+  @Specialization
+  Object doLong(long value) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().number().getInteger();
+  }
+
+  @Specialization
+  Object doBigInteger(EnsoBigInteger value) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().number().getInteger();
+  }
+
+  @Specialization(guards = {"interop.isDate(value)", "interop.isTime(value)"})
+  Object doDateTime(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().dateTime();
+  }
+
+  @Specialization(
+      guards = {"!interop.isDate(value)", "!interop.isTime(value)", "interop.isTimeZone(value)"})
+  Object doTimeZone(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().timeZone();
+  }
+
+  @Specialization(guards = {"interop.isDate(value)", "!interop.isTime(value)"})
+  Object doDate(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().date();
+  }
+
+  @Specialization(guards = {"!interop.isDate(value)", "interop.isTime(value)"})
+  Object doTime(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().timeOfDay();
+  }
+
+  @Specialization(
+      guards = {
+        "interop.hasMetaObject(value)",
+        "!types.hasType(value)",
+        "!interop.isDate(value)",
+        "!interop.isTime(value)",
+        "!interop.isTimeZone(value)"
+      })
+  Object doMetaObject(
+      Object value,
+      @CachedLibrary(limit = "3") InteropLibrary interop,
+      @CachedLibrary(limit = "3") TypesLibrary types) {
+    try {
+      return interop.getMetaObject(value);
+    } catch (UnsupportedMessageException e) {
+      CompilerDirectives.transferToInterpreter();
+      Builtins builtins = Context.get(this).getBuiltins();
+      throw new PanicException(builtins.error().makeCompileError("invalid meta object"), this);
+    }
+  }
+
+  @Specialization(guards = {"types.hasType(value)", "!interop.isNumber(value)"})
```

Comment:
Why is this `!isNumber` check needed? Aren't specializations resolved from top-to-bottom so that the numeric checks above should always take precedence??

Reply:
It is needed. It's a common misconception to think of it as a simple top-to-bottom (one I hit myself a few times in the past). It's because e.g. long and double values are also Objects that have type, except not the one we want. And once it specializes for `Ref` it will keep on using that specialization.
Consider 
```
from Standard.Base import all
import Standard.Base.Runtime.Ref
polyglot java import java.lang.System

main =
    r = Ref.new 1
    tpe_1 = Meta.type_of r
    tpe_2 = Meta.type_of 2

    System.out.println tpe_1
    System.out.println tpe_2
```
without the additional check, it would return 
```
Ref
Small_Integer
```
rather than the intended
```
Ref
Integer
```
Here is a more elaborate example https://www.stefan-marr.de/2018/04/a-truffle-specialization-pitfall/